Add support for ReflectionMetadata

JohnnyMorganz opened the pull request:
Downloads and parses ReflectionMetadata (credit to roblox-api-explorer) and then ~~injects the summaries as descriptions~~ creates described classes and enums

Also updates engine and dependencies - primarily to make use of Typescript optional chaining (Closes #35)

Closes #30

JohnnyMorganz committed:
Update engine and dependencies

JohnnyMorganz committed:
Add ReflectionMetadata support

JohnnyMorganz committed:
Update CHANGELOG.md

Kampfkarren commented on src/dump.ts:
```diff
@@ -145,10 +147,57 @@ export interface ApiDump {
+              i.Properties[0].string.find((s: any) => s.$.name === "Name")?._,
+              i.Properties[0].string.find((s: any) => s.$.name === "summary")?._,
+            ])
+            .filter((entry: any) => entry.length === 2),
```
Replace this and all other anys.

Kampfkarren commented on src/dump.ts:
```diff
@@ -145,10 +147,57 @@ export interface ApiDump {
+        (s: any) => s.$.name === "summary",
+      )?._
+
+      if (entry && entry.Item) {
```
Explicit `!== undefined` checks

Kampfkarren commented on src/dump.ts:
```diff
@@ -145,10 +147,57 @@ export interface ApiDump {
+/*
+    Code borrowed from https://github.com/evaera/vscode-roblox-api-explorer/blob/master/src/api.ts
+*/
+async function injectDescriptions(classes: ApiClass[]) {
```
I'm not really a fan of functions like this that mutate over their properties. Can you make this treat `classes` as immutable and return a new version?

JohnnyMorganz committed:
Cleanup and add descriptions for Enums

JohnnyMorganz commented:
I have updated the code for the changes, and also added support for loading Enum descriptions from the RMD.
I am quite certain that the code now creates copies of Classes and Enums respectively, and then returns new objects, however another look at that by somebody else wouldn't hurt.

JohnnyMorganz committed:
Cleanup

Kampfkarren commented on src/dump.ts:
```diff
@@ -147,57 +150,166 @@ export interface ApiDump {
+            }
+        }
+
+        const describedClass: ApiClass = Object.assign({}, classEntry,
+            {
+                Description: summary,
+                Members: describedMembers,
+            })
+        describedClasses.push(describedClass)
```
This should be equivalent to:

```suggestion
        describedClasses.push({
                Description: summary,
                Members: describedMembers,
                ...classEntry
        })
```

JohnnyMorganz replied:
Yep, I'll change that (except I believe the spread has to go before the overwritten properties)

Kampfkarren replied:
You're right, yeah.

Kampfkarren commented on src/dump.ts:
```diff
@@ -147,57 +150,166 @@ export interface ApiDump {
     }))
-    await injectDescriptions(dump.Classes)
-    return dump
+    const describedDump = await createDescribedDump(dump)
+    return describedDump
```
```suggestion
    return await createDescripedDump(dump)
```

Kampfkarren commented on src/dump.ts:
```diff
@@ -147,57 +150,166 @@ export interface ApiDump {
+    for (const enumEntry of enums) {
+        const entry = rmd.roblox.Item.find(
+            (i: RMDParentItem) => i.$.class === "ReflectionMetadataEnums",
+        )?.Item.find((i: RMDItem) =>
```
Is the type here necessary? `Item` is already defined as `RMDItem[]`. Check for other unnecessary types as well.

Kampfkarren commented on src/dump.ts:
```diff
@@ -147,57 +150,166 @@ export interface ApiDump {
+            (s: RMDStringProperty) => s.$.name === "summary",
+        )?._
+
+        const describedEnumItems: ApiEnumItem[] = ([] as ApiEnumItem[]).concat(enumEntry.Items)
```
```suggestion
        const describedEnumItems = [...enumEntry.Items]
```

Kampfkarren commented on src/dump.ts:
```diff
@@ -3,8 +3,10 @@
 */
 import * as request from "request-promise-native"
 import * as vscode from "vscode"
+import { parseStringPromise } from "xml2js"
```
Alphabetical order

JohnnyMorganz replied:
Would it not be better separating the "whole package" exports from the specific named exports? (See extension.ts for example)

Kampfkarren replied:
I forgot that JavaScript imports are always ugly, been working in a lot of other stuff recently. Never mind this.

Kampfkarren commented on src/dump.ts:
```diff
@@ -145,10 +149,166 @@ export interface ApiDump {
+    },
+    Properties: RMDPropertyHolder[],
+    Item?: RMDItem[],
+
```
Remove this random new line

Kampfkarren commented:
Looking better! Just some style qualms now.

JohnnyMorganz committed:
Cleanup

Kampfkarren commented on src/dump.ts:
```diff
@@ -191,29 +190,29 @@ function createDescribedClasses(rmd: RMDump, classes: ApiClass[]): ApiClass[] {
 
         const summary = entry?.Properties[0].string.find(
-            (s: RMDStringProperty) => s.$.name === "summary",
+            s => s.$.name === "summary",
```
Didn't notice this before, please try to avoid shortened variable names. You don't need to be extremely verbose, but for example if you were iterating over players, the variable would be `player`--not `ply` or `p`.

JohnnyMorganz committed:
Fix variable names

Kampfkarren reviewed (approved):
Will test when I can, thanks!

Kampfkarren commented:
I'm testing it, but I'm not really sure what I'm supposed to be looking for to know if it's working.

JohnnyMorganz commented:
There isn't much to explicitly test in this, it actually just populates the "Description" properties for classes and Enums, which can later be added on to documentation strings for completion items. I guess you could just test it by exporting the updated JSON produced and checking the first few items to see if it is suitable?

JohnnyMorganz committed:
Merge branch 'master' into reflection

Kampfkarren commented:
I'm still not sure what I'm supposed to be looking for here. Are these supposed to have descriptions?

![image](https://user-images.githubusercontent.com/3190756/89270699-96609200-d5f0-11ea-8916-a98e93a6962b.png)

JohnnyMorganz committed:
Show description in documentation

JohnnyMorganz commented:
As #32 has been merged, I've pushed a commit to now use the Description property in the documentation strings.
![image](https://user-images.githubusercontent.com/19635171/89283523-36e0a180-d645-11ea-9592-ea0f30cc4733.png)
Note, some properties and Enums don't have descriptions in the RMD so nothing will show up.

For some services, the inheritance tree is quite large and so the documentation string is quite large - I have this resolved in another (upcoming) PR, however if you want me to move that over here I can do

Kampfkarren commented on src/enum.ts:
```diff
@@ -14,19 +14,24 @@ export class EnumCompletionProvider implements vscode.CompletionItemProvider {
-                enumNamesAndItems[eenum.Name] = eenum.Items.map(
-                    (item) => new vscode.CompletionItem(item.Name, vscode.CompletionItemKind.EnumMember))
+                enumNamesAndItems[eenum.Name] = eenum.Items.map(item => {
+                        const completionItem = new vscode.CompletionItem(item.Name,
+                            vscode.CompletionItemKind.EnumMember)
+                        completionItem.documentation = new vscode.MarkdownString(`${item.Description ? item.Description + "\n\n" : ""}[Developer Reference](https://developer.roblox.com/en-us/api-reference/enum/${eenum.Name})`)
+                        return completionItem
+                    })
```
De-indent this one.

JohnnyMorganz committed:
Cleanup

JohnnyMorganz committed:
Merge branch 'master' into reflection

Kampfkarren commented on CHANGELOG.md:
```diff
@@ -8,6 +8,7 @@ Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how
 - Fixed properties of services not showing up in autofill.
 - Improved documentation provided in autofill, and added relevant Developer Reference links.
 - Added color picker support to Color3.fromHSV().
+- Added support for ReflectionMetadata. This now downloads the file and injects descriptions to be used
```
```suggestion
- Added support for ReflectionMetadata. This now downloads the file and injects descriptions to be used.
```

Kampfkarren committed:
Update CHANGELOG.md
